### ISSUE TITLE
question: How to do conditional useQuery?

### ISSUE THREAD

**Issue by melounek** [relay-tools/relay-hooks#57]
Hi, 
any idea, how to "elegantly" solve this one?
```javascript
  const { deckId } = useMyContext(); // <- this can be null during refreshing page
  if (!deckId) return null;
  const { props } = useQuery<cardsQuery>(query, { deckId }, {});
```
I'm obviously using conditions with hooks which will and up with the error: `React Hook "useQuery" is called conditionally. React Hooks must be called in the exact same order in every component render`.

The only way to go here is to create one more wrapping component, which will conditionally render this component. But then `useQuery` looks like I need even more complicated code then with `QueryRenderer`

What came into my mind as a solution is adding option `supressQuery?: boolean` which will allow me conditionally "turn off" the query.

**Comment by morrys** [comment]
this issue is currently a problem also detected in the official repository: https://github.com/facebook/relay/issues/2936

Introducing a new property "supressQuery" or "skip" is a possible solution, but before introducing features not present in "relay", an official answer from the relay team would be useful.

**Comment by melounek** [comment]
Thanks for the answer.

And there is one approach I will be using for now.
```javascript
const { deckId } = useMyContext();
const { props } = useQuery<cardsQuery>(
    query,
    { deckId: String(deckId) },
    { fetchPolicy: !deckId ? "store-only" : "store-or-network" }
  );
```

The `store-only` will prevent sending the query to server at least and it will always return `null` which is exactly what I need.

So for me that's solved and feel free to reopen this issue if needed.

**Comment by morrys** [comment]
I'm considering making the query parameter optional to avoid running the query.
This modification as the issue https://github.com/relay-tools/relay-hooks/issues/62 not match with relay-experimental

would you like as an alternative solution for this problem? @sibelius @stramel @melounek

**Comment by melounek** [comment]
sounds good

**Comment by stramel** [comment]
Hmmm I gave this some thought, here are some options that I had thought of and are based on stuff I have read.

**1) Optional Query:**

Simplest option but not as intuitive of how to accomplish it. Also doesn't work well for more advanced cases

```js
const skip = true // use your condition here
useQuery(!skip ? query : undefined, variables, options)
```

**2) Skip/Surpess Option:**

This option is similar to apollo's `skip` option. Great for basic cases and can be used in more advanced cases but may require a lot of workarounds.
https://www.apollographql.com/docs/react/data/queries/#options

```js
const skip = true // use your condition here
useQuery(query, variables, { ...options, skip })
```

**3) Manual Trigger / Lazy Query:**

This option is similar to apollo's `useLazyQuery`. This requires a bit more code for a basic case but allows greater control of the overall fetching and advanced use-cases. https://www.apollographql.com/docs/react/data/queries/#executing-queries-manually

I haven't tried playing around with `useLazyLoadQuery` yet and am not sure if that covers this case just in a different manner. 

```js
const [fetchQuery, { props, error }] = useLazyQuery(query)

useEffect(() => {
  const skip = true // use your condition here
  if (!skip) {
    fetchQuery(variables, options)
  }
}, [])
```

**4) Leave the API alone:**

Even less intuitive, feels almost hacky. Still may not work if the data exists in the store.

```js
const skip = true // use your condition here
useQuery(query, variables, { ...options, fetchPolicy: skip ? "store-only" : "store-or-network" })
```

Personally, I like supporting both 2 and 3. This is similar to apollo's setup and allows for easily handling any scenario. 

I don't like option 1, the optional query, because it's counter-intuitive. The query isn't really optional just the execution of the query. I guess you could look at it as a single unit but maybe it's just my opinion.

I definitely don't like option 4, it feels like a workaround for the lack of a solution. 

Hope this is helpful @morrys!

**Comment by morrys** [comment]
Excellent summary and thanks for your opinion :)
With these considerations, solution 2 is the one with the least impact and allows excellent flexibility.

**Comment by morrys** [comment]
I have just released version 3.2.0
https://github.com/relay-tools/relay-hooks/releases/tag/v3.2.0

Feel free to open new issues if you find problems :)
Thanks
@stramel @melounek @roman-khasimyanov
